### PR TITLE
add task profile APIs

### PR DESCRIPTION
**Pull request by Zhangxunmt**
### Description
Add ML-Common Profile APIs to return runtime information for ML Models/Tasks. 
Will add model profile, per node profile, and more unit tests in the next revision.
 
### Issues Resolved
https://github.com/opensearch-project/ml-commons/issues/410
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        super.writeTo(out);
+        if (mlNodeTasks != null) {
+            out.writeBoolean(true);
+            out.writeMap(mlNodeTasks, StreamOutput::writeString, (o, r) -> r.writeTo(o));
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java:58]: Can we add javadoc for the public functions?

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     * @throws IOException throws an IO exception if the StreamInput cannot be reML from
+     */
+    public MLTaskProfileNodeResponse(StreamInput in) throws IOException {
+        super(in);
+        if (in.readBoolean()) {
+            this.mlNodeTasks = in.readMap(StreamInput::readString, MLTask::new);
+        }
+    }
+
+    public MLTaskProfileNodeResponse(DiscoveryNode node, Map<String, MLTask> nodeTasks) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java:38]: Map<String, MLTask> --> is the string for node task id? I hope we aren't expecting readers to have intensive knowledge of this package.

**Review by dhrubo-os — CHANGES_REQUESTED** [review]
Overall I would suggest to add more documentation throughout the PR. As a new reader of this package it was not quite easy for me to understand what's going on.

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+package org.opensearch.ml.action.profile;
+
+import org.opensearch.action.ActionType;
+import org.opensearch.ml.constant.CommonValue;
+
+public class MLTaskProfileAction extends ActionType<MLTaskProfileResponse> {
+    // Internal Action which is not used for public facing RestAPIs.
+    public static final String NAME = CommonValue.ACTION_PREFIX + "tasks/profile";
+    public static final MLTaskProfileAction INSTANCE = new MLTaskProfileAction();
```
[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java:14]: Check the other Actions. These are normal names.

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    private Map<String, MLTask> mlNodeTasks;
+    /**
+     * Constructor
+     *
+     * @param in StreamInput
+     * @throws IOException throws an IO exception if the StreamInput cannot be reML from
+     */
+    public MLTaskProfileNodeResponse(StreamInput in) throws IOException {
+        super(in);
+        if (in.readBoolean()) {
```
[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java:33]: This correspond to line 51 out.writeBoolean(true) if mlNodeTasks != null.

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java (hunk 2 above)
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileNodeResponse.java:38]: I added more comments in the next revision. The key is the taskId and the value is the MLTask that we want to retrieve in the memory to the users.

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java:3]: Yes this class is actually only a hooker that defines the Response and the real action is in the MLTaskProfileTransportAction.

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java:14]: what do you mean by normal names? Is this a convention to keep this way? Is is name of the action? ACTION_NAME?

File: plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+package org.opensearch.ml.action.profile;
+
+import org.opensearch.action.ActionType;
+import org.opensearch.ml.constant.CommonValue;
+
+public class MLTaskProfileAction extends ActionType<MLTaskProfileResponse> {
+    // Internal Action which is not used for public facing RestAPIs.
+    public static final String NAME = CommonValue.ACTION_PREFIX + "tasks/profile";
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/profile/MLTaskProfileAction.java:13]: How about "tasks/profile" -> "profile" as we are going to profile not just for task, also for model and may add other things in future.

**Comment by ylwu-amzn** [comment]
From the description "Will add model profile, per node profile, and more unit tests in the next revision.", seems this PR only has task profile. Are you planning to add model profile in this PR or next?

File: plugin/src/main/java/org/opensearch/ml/rest/RestMLTaskProfileAction.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import org.opensearch.rest.BaseRestHandler;
+import org.opensearch.rest.BytesRestResponse;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.rest.RestStatus;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+@Log4j2
+public class RestMLTaskProfileAction extends BaseRestHandler {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/rest/RestMLTaskProfileAction.java:37]: Going to create multiple profile APIs, like one for task profile, another one for model? How about we keep one profile API and user can specify the target they want to detect?

Comment (Zhangxunmt) [reply plugin/src/main/java/org/opensearch/ml/rest/RestMLTaskProfileAction.java:37]: I understand your point. We can refactor this one and let user to define the target in the request body.

**Pull request closed without merging**